File: mac/Sources/Fluent/AppModel.swift
import AppKit
import AVFoundation
import FluentCore
import FluentMacKit
import Observation
import SwiftUI

/// App-wide state for the UI. Settings are mirrored into observable properties and written
/// straight back to UserDefaults, like the iPhone app's `AppModel`.
@MainActor
@Observable
final class AppModel {
    enum Tab: String, CaseIterable, Identifiable {
        case dictate, history, style, settings
        var id: String { rawValue }
        var title: String { rawValue.capitalized }
        var symbol: String {
            switch self {
            case .dictate: "mic.fill"
            case .history: "clock"
            case .style: "textformat"
            case .settings: "gearshape"
            }
        }
    }

    let store: FluentCore.Settings
    let mac: MacSettings
    let historyStore: HistoryStore
    let dictation: DictationController
    /// Set when launched with `--demo`: nothing is saved to the real settings.
    let demo: String?

    var tab: Tab = .dictate
    var termsAccepted: Bool
    var setupComplete: Bool { didSet { store.setupComplete = setupComplete } }
    var themeID: String { didSet { store.themeID = themeID } }
    var mode: TranscriptionMode { didSet { store.mode = mode } }
    var languageCode: String { didSet { store.languageCode = languageCode } }
    var vocabulary: [String] { didSet { store.vocabulary = vocabulary } }
    var historyEnabled: Bool { didSet { store.historyEnabled = historyEnabled } }
    var styleEnabled: Bool { didSet { store.styleEnabled = styleEnabled } }
    private(set) var styles: [StyleCategory: WritingStyle] = [:]

    var bubbleEnabled: Bool { didSet { mac.bubbleEnabled = bubbleEnabled } }
    var bubbleSize: Double { didSet { mac.bubbleSize = bubbleSize } }
    var bubbleOpacity: Double { didSet { mac.bubbleOpacity = bubbleOpacity } }
    var excludedApps: [String] { didSet { mac.excludedApps = excludedApps } }
    var snoozeChoice: SnoozeChoice { didSet { mac.snoozeChoice = snoozeChoice } }
    private(set) var snoozeUntil: Double
    var holdKey: HoldKey { didSet { mac.holdKey = holdKey; onShortcutsChanged?() } }
    var toggleShortcut: ToggleShortcut { didSet { mac.toggleShortcut = toggleShortcut; onShortcutsChanged?() } }
    var overlayThemeID: String { didSet { mac.overlayThemeID = overlayThemeID } }
    var soundsEnabled: Bool { didSet { mac.soundsEnabled = soundsEnabled } }
    private(set) var launchAtLogin: Bool
    var launchAtLoginError: String?

    private(set) var hasApiKey: Bool
    private(set) var maskedKey = ""
    private(set) var history: [Transcript] = []

    /// Live permission state, re-read every second while Fluent runs.
    private(set) var micStatus: AVAuthorizationStatus = .notDetermined
    private(set) var axTrusted = false

    var onShortcutsChanged: (() -> Void)?

    var palette: Palette { Palette.byID(themeID) }
    var overlayPalette: Palette {
        overlayThemeID == MacSettings.matchApp ? palette : Palette.byID(overlayThemeID)
    }
    var micGranted: Bool { micStatus == .authorized }
    var snoozed: Bool { snoozeUntil == -1 || snoozeUntil > Date().timeIntervalSince1970 }
    var ready: Bool { micGranted && axTrusted && hasApiKey }

    init(defaults: UserDefaults = .standard, demo: String? = nil) {
        self.demo = demo
        store = FluentCore.Settings(defaults: defaults)
        mac = MacSettings(defaults: defaults)
        let support = demo == nil
            ? FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0].appendingPathComponent("Fluent")
            : FileManager.default.temporaryDirectory.appendingPathComponent("Fluent-demo-\(UUID().uuidString)")
        historyStore = HistoryStore(directory: support)
        mac.clearRestartSnooze()

        termsAccepted = store.termsAccepted
        setupComplete = store.setupComplete
        themeID = Palette.byID(store.themeID).id
        mode = store.mode
        languageCode = store.languageCode
        vocabulary = store.vocabulary
        historyEnabled = store.historyEnabled
        styleEnabled = store.styleEnabled
        bubbleEnabled = mac.bubbleEnabled
        bubbleSize = mac.bubbleSize
        bubbleOpacity = mac.bubbleOpacity
        excludedApps = mac.excludedApps
        snoozeChoice = mac.snoozeChoice
        snoozeUntil = mac.snoozeUntil
        holdKey = mac.holdKey
        toggleShortcut = mac.toggleShortcut
        overlayThemeID = mac.overlayThemeID
        soundsEnabled = mac.soundsEnabled
        launchAtLogin = demo == nil ? LaunchAtLogin.enabled : false
        let key = demo == nil ? KeychainStore.load() : nil
        hasApiKey = key?.isEmpty == false
        maskedKey = Self.mask(key)
        dictation = DictationController(settings: store)
        for c in StyleCategory.allCases { styles[c] = store.style(for: c) }
        history = historyStore.all()
        dictation.model = self
        refreshPermissions()
    }

    // MARK: permissions

    /// Demo screenshots of a set-up Mac: shows the app as it looks once permissions and a key
    /// are in place. Only `--demo` uses it; the CI runner has no microphone or key of its own.
    private var permissionsFrozen = false

    func pretendReady() {
        permissionsFrozen = true
        micStatus = .authorized
        axTrusted = true
        hasApiKey = true
        maskedKey = "AIza••••••••x9Qk"
    }

    func refreshPermissions() {
        if permissionsFrozen { return }
        micStatus = Recorder.permission
        axTrusted = AccessibilityAccess.trusted
        if snoozeUntil > 0, !snoozed { snoozeUntil = 0; mac.clearSnooze() }
    }

    func requestMicrophone() {
        if micStatus == .notDetermined {
            Task {
                _ = await Recorder.requestPermission()
                refreshPermissions()
            }
        } else {
            SystemSettings.microphone()
        }
    }

    func requestAccessibility() {
        AccessibilityAccess.request()
        SystemSettings.accessibility()
    }

    // MARK: terms, key

    func acceptTerms() {
        store.acceptTerms()
        termsAccepted = true
    }

    func saveApiKey(_ key: String) {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { KeychainStore.delete() } else { KeychainStore.save(trimmed) }
        hasApiKey = !trimmed.isEmpty
        maskedKey = Self.mask(trimmed)
    }

    func testApiKey() async -> Result<String, TranscriptionError> {
        await GeminiClient(apiKey: KeychainStore.load() ?? "").testConnection()
    }

    private static func mask(_ key: String?) -> String {
        guard let key, key.count > 8 else { return key == nil || key!.isEmpty ? "" : "••••" }
        return String(key.prefix(4)) + String(repeating: "•", count: 8) + String(key.suffix(4))
    }

    // MARK: style

    func style(for c: StyleCategory) -> WritingStyle { styles[c] ?? c.defaultStyle }

    func setStyle(_ s: WritingStyle, for c: StyleCategory) {
        store.setStyle(s, for: c)
        styles[c] = store.style(for: c)
    }

    // MARK: snooze, login

    func snooze(_ choice: SnoozeChoice? = nil) {
        mac.snooze(choice ?? snoozeChoice)
        snoozeUntil = mac.snoozeUntil
    }

    func resume() {
        mac.clearSnooze()
        snoozeUntil = 0
    }

    var snoozeLabel: String {
        if snoozeUntil == -1 { return "Snoozed until Fluent restarts" }
        let until = Date(timeIntervalSince1970: snoozeUntil)
        return "Snoozed until \(until.formatted(date: .omitted, time: .shortened))"
    }

    func setLaunchAtLogin(_ on: Bool) {
        launchAtLoginError = LaunchAtLogin.set(on)
        launchAtLogin = LaunchAtLogin.enabled
        mac.launchAtLogin = launchAtLogin
    }

    func exclude(_ bundleID: String) {
        if !excludedApps.contains(bundleID) { excludedApps.append(bundleID) }
    }

    // MARK: history

    func record(_ t: Transcript) {
        guard historyEnabled else { return }
        historyStore.add(t)
        history.insert(t, at: 0)
    }

    func delete(_ t: Transcript) {
        historyStore.delete(t.id)
        history.removeAll { $0.id == t.id }
    }

    func clearHistory() {
        historyStore.clear()
        history = []
    }

    var wordsToday: Int {
        history.filter { Calendar.current.isDateInToday($0.createdAt) }.reduce(0) { $0 + $1.wordCount }
    }

    var dictationsToday: Int {
        history.filter { Calendar.current.isDateInToday($0.createdAt) }.count
    }

    /// Only used by `--demo history`, so the History screenshot has something to show.
    func seedDemoHistory() {
        let now = Date()
        let samples = [
            ("Hey team, the new build is up. Can everyone test the onboarding flow before Friday?", 0.0, 6.2),
            ("Hi Sam,\n\nThanks for the notes. I'll send the updated deck tomorrow morning.\n\nBest,\nHammaad", 3600.0, 8.4),
            ("Pick up oat milk, coffee beans and something for dinner", 7200.0, 3.1),
            ("Remind me to book the dentist for next week", 90_000.0, 2.7),
        ]
        for (text, ago, secs) in samples.reversed() {
            let t = Transcript(text: text, createdAt: now.addingTimeInterval(-ago), durationSeconds: secs)
            historyStore.add(t)
        }
        history = historyStore.all()
    }
}

File: src/Fluent/AppModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Fluent.Core;

namespace Fluent;

public abstract class Observable : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;
    protected void Raise([CallerMemberName] string? name = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    protected void RaiseAll() => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(null));

    protected bool Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        Raise(name);
        return true;
    }
}

public enum Tab { Dictate, History, Style, Settings }

/// <summary>App-wide state for the UI. Settings are mirrored into observable properties and written
/// straight back to settings.json, like the Mac's <c>AppModel</c>.</summary>
public sealed class AppModel : Observable
{
    public Settings Store { get; }
    public HistoryStore HistoryStore { get; }
    public DictationController Dictation { get; }
    /// <summary>Set when launched with --demo: nothing touches the real settings or key.</summary>
    public string? Demo { get; }

    public event Action? ThemeChanged;
    public event Action? ShortcutsChanged;
    public event Action? TabChanged;

    public AppModel(Settings store, string dataDir, string? demo = null)
    {
        Store = store;
        Demo = demo;
        HistoryStore = new HistoryStore(dataDir);
        Store.ClearRestartSnooze();
        var key = demo is null ? KeyStore.Load() : null;
        hasApiKey = !string.IsNullOrEmpty(key);
        maskedKey = Mask(key);
        launchAtLogin = demo is null && LaunchAtLogin.Enabled;
        foreach (var t in HistoryStore.All()) History.Add(t);
        Dictation = new DictationController(this);
        RefreshPermissions();
    }

    Tab tab = Tab.Dictate;
    public Tab Tab { get => tab; set { if (Set(ref tab, value)) TabChanged?.Invoke(); } }

    public bool TermsAccepted => Store.TermsAccepted;
    public void AcceptTerms() { Store.AcceptTerms(); Raise(nameof(TermsAccepted)); }

    public bool SetupComplete { get => Store.SetupComplete; set { Store.SetupComplete = value; Raise(); } }

    public string ThemeId
    {
        get => Palette.ById(Store.ThemeId).Id;
        set { Store.ThemeId = value; Raise(); Raise(nameof(Palette)); Raise(nameof(OverlayPalette)); ThemeChanged?.Invoke(); }
    }
    public Palette Palette => Palette.ById(Store.ThemeId);
    public string OverlayThemeId
    {
        get => Store.OverlayThemeId;
        set { Store.OverlayThemeId = value; Raise(); Raise(nameof(OverlayPalette)); }
    }
    public Palette OverlayPalette => OverlayThemeId == Settings.MatchApp ? Palette : Palette.ById(OverlayThemeId);

    public TranscriptionMode Mode { get => Store.Mode; set { Store.Mode = value; Raise(); Raise(nameof(ModeDetail)); Raise(nameof(StyleNote)); } }
    public string ModeDetail => Mode.Detail();
    public string LanguageCode { get => Store.LanguageCode; set { Store.LanguageCode = value; Raise(); } }

    public ObservableCollection<string> Vocabulary { get; } = [];
    public void AddWord(string w)
    {
        w = w.Trim();
        var list = Store.Vocabulary.ToList();
        if (w.Length == 0 || list.Contains(w)) return;
        list.Add(w);
        Store.Vocabulary = list;
        Vocabulary.Add(w);
    }
    public void RemoveWord(string w)
    {
        Store.Vocabulary = Store.Vocabulary.Where(x => x != w).ToList();
        Vocabulary.Remove(w);
    }

    public bool HistoryEnabled { get => Store.HistoryEnabled; set { Store.HistoryEnabled = value; Raise(); RaiseStats(); } }
    public bool StyleEnabled { get => Store.StyleEnabled; set { Store.StyleEnabled = value; Raise(); } }
    public string StyleNote => Mode == TranscriptionMode.Verbatim ? "Off in Verbatim mode, which keeps every word exactly" : "Applies in Smart mode";
    public WritingStyle StyleFor(StyleCategory c) => Store.StyleFor(c);
    public void SetStyle(WritingStyle s, StyleCategory c) { Store.SetStyle(s, c); Raise("Styles"); }

    public bool BubbleEnabled { get => Store.BubbleEnabled; set { Store.BubbleEnabled = value; Raise(); } }
    public double BubbleSize { get => Store.BubbleSize; set { Store.BubbleSize = value; Raise(); } }
    public double BubbleOpacity { get => Store.BubbleOpacity; set { Store.BubbleOpacity = value; Raise(); } }

    public ObservableCollection<string> ExcludedApps { get; } = [];
    public void Exclude(string process)
    {
        var p = AppCategories.Normalize(process);
        if (p.Length == 0 || Store.ExcludedApps.Contains(p)) return;
        Store.ExcludedApps = [.. Store.ExcludedApps, p];
        ExcludedApps.Add(p);
    }
    public void Include(string process)
    {
        Store.ExcludedApps = Store.ExcludedApps.Where(x => x != process).ToList();
        ExcludedApps.Remove(process);
    }

    public SnoozeChoice SnoozeChoice { get => Store.SnoozeChoice; set { Store.SnoozeChoice = value; Raise(); } }
    public bool Snoozed => Store.SnoozeActive();
    public string SnoozeLabel => Store.SnoozeUntil == -1 ? "Snoozed until Fluent restarts"
        : $"Snoozed until {DateTimeOffset.FromUnixTimeMilliseconds((long)(Store.SnoozeUntil * 1000)).ToLocalTime():t}";
    public void Snooze(SnoozeChoice? c = null) { Store.Snooze(c ?? SnoozeChoice); RaiseSnooze(); }
    public void Resume() { Store.ClearSnooze(); RaiseSnooze(); }
    void RaiseSnooze() { Raise(nameof(Snoozed)); Raise(nameof(SnoozeLabel)); Raise(nameof(Status)); Raise(nameof(Headline)); }

    public HoldKey HoldKey { get => Store.HoldKey; set { Store.HoldKey = value; Raise(); RaiseShortcuts(); } }
    public ToggleShortcut ToggleShortcut { get => Store.ToggleShortcut; set { Store.ToggleShortcut = value; Raise(); RaiseShortcuts(); } }
    public string ShortcutSummary => $"Hold {HoldKey.Label()} to talk · {ToggleShortcut.Label} to start and stop";
    void RaiseShortcuts() { Raise(nameof(ShortcutSummary)); ShortcutsChanged?.Invoke(); }

    public bool SoundsEnabled { get => Store.SoundsEnabled; set { Store.SoundsEnabled = value; Raise(); } }

    bool launchAtLogin;
    public bool LaunchAtLoginEnabled
    {
        get => launchAtLogin;
        set
        {
            LaunchAtLoginError = Demo is null ? LaunchAtLogin.Set(value) : null;
            launchAtLogin = Demo is null ? LaunchAtLogin.Enabled : value;
            Raise(); Raise(nameof(LaunchAtLoginError));
        }
    }
    public string? LaunchAtLoginError { get; private set; }

    // permissions and key

    Recorder.MicState mic = Recorder.MicState.Allowed;
    public Recorder.MicState Mic { get => mic; private set { if (Set(ref mic, value)) { Raise(nameof(MicAllowed)); Raise(nameof(Ready)); Raise(nameof(Status)); } } }
    public bool MicAllowed => Mic == Recorder.MicState.Allowed;
    bool frozen;

    /// <summary>Demo screenshots of a set-up PC: the CI machine has no microphone or key of its own.</summary>
    public void PretendReady()
    {
        frozen = true;
        mic = Recorder.MicState.Allowed;
        hasApiKey = true;
        maskedKey = "AIza••••••••x9Qk";
        RaiseAll();
    }

    public void RefreshPermissions()
    {
        if (frozen) return;
        Mic = Recorder.Check();
        if (Store.SnoozeUntil > 0 && !Snoozed) { Store.ClearSnooze(); RaiseSnooze(); }
    }

    bool hasApiKey;
    string maskedKey;
    public bool HasApiKey => hasApiKey;
    public string MaskedKey => maskedKey;
    public bool Ready => MicAllowed && HasApiKey;

    public void SaveApiKey(string key)
    {
        var trimmed = key.Trim();
        if (Demo is null) { if (trimmed.Length == 0) KeyStore.Delete(); else KeyStore.Save(trimmed); }
        hasApiKey = trimmed.Length > 0;
        maskedKey = Mask(trimmed);
        Raise(nameof(HasApiKey)); Raise(nameof(MaskedKey)); Raise(nameof(Ready)); Raise(nameof(Status));
    }

    public string? ApiKey => Demo is null ? KeyStore.Load() : null;

    public Task<TranscriptionResult> TestApiKeyAsync() => new GeminiClient(ApiKey ?? "").TestConnectionAsync();

    static string Mask(string? key)
    {
        if (string.IsNullOrEmpty(key)) return "";
        if (key.Length <= 8) return "••••";
        return key[..4] + new string('•', 8) + key[^4..];
    }

    public (string Text, string Kind) Status =>
        !TermsAccepted ? ("Open Fluent to get started", "warn")
        : Mic == Recorder.MicState.BlockedByWindows ? ("Microphone blocked", "danger")
        : Mic == Recorder.MicState.NoDevice ? ("No microphone", "danger")
        : !HasApiKey ? ("API key needed", "warn")
        : Snoozed ? ("Snoozed", "faint")
        : ("Ready", "ok");

    public string Headline => Dictation.Phase switch
    {
        DictationPhase.Recording => "Listening.",
        DictationPhase.Paused => "Paused.",
        DictationPhase.Transcribing => "Writing it down.",
        _ => Snoozed ? "Snoozed." : "Ready when you are.",
    };

    // history

    public ObservableCollection<Transcript> History { get; } = [];

    public void Record(Transcript t)
    {
        if (!HistoryEnabled) return;
        HistoryStore.Add(t);
        History.Insert(0, t);
        RaiseStats();
    }

    public void Delete(Transcript t)
    {
        HistoryStore.Delete(t.Id);
        History.Remove(t);
        RaiseStats();
    }

    public void ClearHistory()
    {
        HistoryStore.Clear();
        History.Clear();
        RaiseStats();
    }

    public int WordsToday => History.Where(t => t.CreatedAt.LocalDateTime.Date == DateTime.Today).Sum(t => t.WordCount);
    public int DictationsToday => History.Count(t => t.CreatedAt.LocalDateTime.Date == DateTime.Today);
    void RaiseStats() { Raise(nameof(WordsToday)); Raise(nameof(DictationsToday)); }

    /// <summary>Only used by --demo, so the History screenshot has something to show.</summary>
    public void SeedDemoHistory()
    {
        var now = DateTimeOffset.Now;
        (string, double, double)[] samples =
        [
            ("Hey team, the new build is up. Can everyone test the onboarding flow before Friday?", 0, 6.2),
            ("Hi Sam,\n\nThanks for the notes. I'll send the updated deck tomorrow morning.\n\nBest,\nAlex", 3600, 8.4),
            ("Pick up oat milk, coffee beans and something for dinner", 7200, 3.1),
            ("Remind me to book the dentist for next week", 90_000, 2.7),
        ];
        foreach (var (text, ago, secs) in samples.Reverse())
            HistoryStore.Add(new Transcript(Guid.NewGuid(), text, now.AddSeconds(-ago), secs));
        History.Clear();
        foreach (var t in HistoryStore.All()) History.Add(t);
        RaiseStats();
    }

    public void LoadCollections()
    {
        Vocabulary.Clear();
        foreach (var w in Store.Vocabulary) Vocabulary.Add(w);
        ExcludedApps.Clear();
        foreach (var p in Store.ExcludedApps) ExcludedApps.Add(p);
    }

    public void NotifyDictation() { Raise(nameof(Headline)); }

    public static string DataDir(string? demo) => demo is null
        ? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Fluent")
        : Path.Combine(Path.GetTempPath(), "Fluent-demo-" + Guid.NewGuid().ToString("N"));
}
